### ISSUE TITLE
plugin v8r-plugin-ndjson does not extend BasePlugin

### ISSUE THREAD

**Issue by wesley-dean-flexion** [chris48s/v8r-plugin-ndjson#1]
Hi!  I'm a huge fan of v8r and I use it on a bunch (hundreds) of projects, especially in conjunction with MegaLinter.

I've been trying to write a plugin to read Markdown, extract YAML front matter, and then parse / validate the YAML using the [v8-plugin-yaml-parser](https://github.com/chris48s/v8r/blob/main/src/plugins/parser-yaml.js) plugin built into v8r.

So, why am I putting this issue here and not into the v8r repo?

Well, I've been trying to track down why a certain error's being thrown:

```
ℹ Loaded config file from .v8rrc.yaml
ℹ Patterns and relative paths will be resolved relative to current working directory: /path/to/repo
✖ Plugin v8r-plugin-markdown-parser does not extend BasePlugin
```

So, I did a search on v8r plugins and came across the repo.  So, I downloaded the index.ts and saved it to `plugins/v8r-plugin-ndjson.mjs` and tried to use it.  I was met with the same error:

```
ℹ Loaded config file from .v8rrc.yaml
ℹ Patterns and relative paths will be resolved relative to current working directory: /path/to/repo
✖ Plugin v8r-plugin-ndjson does not extend BasePlugin
```
I'm using v8r v4.2.1 and my .v8rrc.yaml file reads:

```YAML
---
customCatalog:
  schemas:
    - name: "Job Postings"
      fileMatch: ["pages/jointts/positions/*.md"]
      location: "_schemas/job-posting.json"
        #      parser: "md"

plugins:
  #  - "file:./plugins/v8r-plugin-markdown-parser.mjs"
  - "file:./plugins/v8r-plugin-ndjson.mjs"
```

So, my question to you, @chris48s , is, "what am I doing wrong?  what do I need to do differently with this (ndjson) plugin so that I'll run properly?"  my hope is that by answering that question, I can gain some insight into what I'm doing wrong with the markdown plugin.

Thank you so, so, so much. Being able to include in a pipeline the ability to validate YAML, JSON, etc. against schemae is incredibly helpful to me and I'm deeply appreciative for your hard work.

**Comment by chris48s** [comment]
Hello.

First off, I think you might be the first person other than me who has attempted to write a v8r plugin so it is cool to know that others are starting to find a use for this feature.

I am not able to reproduce the issue you are seeing, but to start off with I've set up a couple of very small example repos:
One where I am pulling in v8r-plugin-ndjson as a package - https://github.com/chris48s/v8r-plugins-example1
and another where I've done what you've done and copied the plugin to a local file - https://github.com/chris48s/v8r-plugins-example2

So you should be able to:

```
git clone https://github.com/chris48s/v8r-plugins-example1.git
cd v8r-plugins-example1
npm ci
npm run validate
```

and

```
git clone https://github.com/chris48s/v8r-plugins-example2.git
cd v8r-plugins-example2
npm ci
npm run validate
```

Both of those work for me. Do they work for you? If not, I guess we start looking at environmental factors.

If those two repos work for you as described, then can you reduce what you're trying to do to a similarly small example that I can run locally and reproduce the behaviour you are seeing?

**Comment by wesley-dean-flexion** [comment]
@chris48s Thank you for your quick response!  That's awesome!!

I'll review what you wrote momentarily; for what it's worth, the following ought to be current at the moment:

https://github.com/GSA-TTS/tts.gsa.gov/blob/v8r-additions/.v8rrc.yaml
https://github.com/GSA-TTS/tts.gsa.gov/blob/v8r-additions/plugins/v8r-plugin-markdown-parser.mjs

Basically, I'm trying to mash together the YAML parser from your repo with the [brainmatter](https://www.npmjs.com/package/brainmatter) module to extract the YAML front matter from a bunch of Markdown files so that v8r can validate the YAML against a [schema](https://github.com/GSA-TTS/tts.gsa.gov/blob/v8r-additions/_schemas/job-posting.json) we put together.

For what it's worth, my background is in DevSecOps and I'm not a Node developer.  It's entirely possible that I'm missing something obvious.  If that's the case, please allow me to apologize in advance.

I'll reread and get back to you shortly.  Thank you again.

**Comment by wesley-dean-flexion** [comment]
I looked at the second example and attempted to replicate it by..

1. moving `v8r-plugin-markdown-parser.mjs` to the repo root
2. changing `.v8rrc.yaml` to reflect the updated path

For what it's worth -- and this is probably an important detail that I neglected to mention previously -- I installed v8r globally (`npm install --global v8r`).

Update

Ok, I removed the global install and installed it locally (`npm install v8r`) and then ran it locally (`node_modules/.bin/v8r -vvvvvvvvv pages/jointts/positions/*.md`) and I see that it's now loading the markdown plugin ( :exclamation: ) and I'm getting what appears to be what's supposed to be displayed:

```
ℹ Processing ./pages/jointts/positions/Senior-software-eng-15-Oct24.md
ℹ Searching for schema in .v8rrc.yaml ...
ℹ Found schema in .v8rrc.yaml ...
ℹ Validating ./pages/jointts/positions/Senior-software-eng-15-Oct24.md against schema from _schemas/job-posting.json ...
✖ require is not defined
```

I'm thinking that the error message there is a result of something we're doing wrong with the schema.

**Comment by wesley-dean-flexion** [comment]
What I'm seeing now is outside of the scope of the original question.  If you want to close this out and I can ask a question less-specific to this plugin in the main v8r repo, that's cool with me.

I'm working from the assumption that the `require is not defined` is something I'm doing wrong somewhere else outside of the scope of this v8r invocation.

I ran MegaLinter locally -- and I know this is likely to be better-answered by the MegaLinter community -- and I'm met with the `does not extend BasePlugin` issue from before:

```
✅ Linted [JSON] files with [v8r]: Found 1 non blocking error(s) - (0.89s)
- Using [v8r v4.2.0] https://megalinter.io/8.3.0/descriptors/json_v8r
- MegaLinter key: [JSON_V8R]
- Rules config: identified by [v8r]
- Number of files analyzed: [3]
- Command: [v8r --ignore-errors _schemas/job-posting.json package-lock.json package.json]
--Error detail:
ℹ Loaded config file from .v8rrc.yml
ℹ Patterns and relative paths will be resolved relative to current working directory: /path/to/file
✖ Plugin v8r-plugin-markdown-parser does not extend BasePlugin
```

This suggests to me that MegaLinter's functioning in a model more like what I was doing previously (i.e., more like a global install of v8r).

So, my new question is if there's a mechanism by which we could use a globally-installed v8r extend into plugins?

For what it's worth, I went through MegaLinter's issue backlog and didn't find anything even remotely related to this.  None of the other linters in MegaLinter have anything like this.

I don't know if there's a setting or a configuration mechanism to get from here to there.  I've no idea if it's even possible.

All of that said, if it's not possible to adapt the plugin mechanism to work with globally-installed v8r, what about..


1. merging the Markdown (well, front matter in Markdown) functionality into v8r proper
2. creating a package for the Markdown extractor plugin so that one could run some variation of `npm install v8r-plugin-markdown-extractor`?  This is probably what you meant with the first example repo..

Tangentially -- and this probably doesn't help or even mean anything to you at all -- I'm working through something similar trying to get yamllint to block builds that have Markdown with invalid YAML in the front matter.

**Comment by chris48s** [comment]
OK, so using your repo, I was able to:

```
git clone https://github.com/GSA-TTS/tts.gsa.gov
cd tts.gsa.gov
git checkout v8r-additions
npm ci
```

Then I edited your v8rrc to put back in

```yaml
plugins:
  - "file:./plugins/v8r-plugin-markdown-parser.mjs"
```

..and then I can run
`./node_modules/.bin/v8r *.md`
or
`npx v8r *.md`

either of which gives me the output

```
ℹ Loaded config file from .v8rrc.yaml
ℹ Patterns and relative paths will be resolved relative to current working directory: /home/chris/dev/v8r-plugins/tts.gsa.gov
Registering md file parser
ℹ Processing ./CONTRIBUTING.md
✖ Could not find a schema to validate ./CONTRIBUTING.md

ℹ Processing ./LICENSE.md
✖ Could not find a schema to validate ./LICENSE.md

ℹ Processing ./overview.md
✖ Could not find a schema to validate ./overview.md

ℹ Processing ./README.md
✖ Could not find a schema to validate ./README.md
```

so yes. Running with a local install is fine.

---

Separately from that, I tried loading a local plugin with a global v8r install (`npm install -g v8r`) and I get the error

```
✖ Cannot find package 'v8r' imported from [path]
```

I haven't managed to reproduce

```
✖ Plugin [plugin name] does not extend BasePlugin
```

yet.

---

In terms of solutions:

I don't particularly have the appetite to take on this feature in core. One of the goals of adding a plugin interface is to allow users to self-serve when it comes to feature requests for more esoteric use-cases.

I'd like to try and work out how to make plugins play nicely with a global install as well as possible if I can. I don't really know what the answer to that is yet, but I have some time off work over Christmas. I'll try and see if I can slot in some time to reproduce the exact error you're seeing and think about possible solutions.

When it comes to more immediate ways to just _get a thing done_

If megalinter give you the ability to install additional NPM packages into the same environment as v8r, then publishing `v8r-plugin-markdown-extractor` as an NPM package might be the quickest way to get round this and get moving. You could use `v8r-plugin-ndjson` as an example of an already published package to see if you can do this before making your own. I've not used megalinter before, but it looks like pre-commands is the feature you're looking for: https://megalinter.io/latest/config-precommands/ The example there shows installing an extra eslint plugin from NPM so hopefully this would work the same way :crossed_fingers: 

Another possibility is to run v8r in CI outside of megalinter. As we've established, the plugin does work with a local install.

**Comment by wesley-dean-flexion** [comment]
> take on this feature in core
Sure, makes perfect sense.  I respect that.

> adding a plugin interface
Yuppers.  I get that, too.  I'm really glad that you've architected v8r to be modular and extensible.

> I'll try and see if I can slot in some time
I appreciate that.  Sincerely.  I value my time away from work when I can get out of my office and work on other projects.  That said, I will likely be non-responsive until mid-January, so please don't miss out on any opportunities to rest, relax, recover, recharge, etc. on my account.

> If megalinter give you the ability to install additional NPM packages
Yes, it can add packages at build time as well as at runtime with precommands, as you mentioned.  I'm confident that it can eventually be made to work.

>  more esoteric use-cases
Yeah, I get that, too.  My attempt to integrate YAML support for front matter into MegaLinter was met with pretty much [the same response](https://github.com/oxsecurity/megalinter/discussions/4066).  But not for the current list of projects I'm on now, I never gave much thought to front matter.  In my reading, I concluded that there is no formal specification for front matter, so it's clearly not a hot topic of discussion many places.

In my current projects -- and none of this is supremely relevant -- I support the CI/CD pipelines between "developers" and production.  A subset of these projects has content managed by non-developers who've been given the task of representing their work as YAML in front matter.  From my (admittedly narrow) perspective, my challenge is to prevent folks from promoting invalid YAML through to the production environments.  Sometimes this is data type mismatches (e.g., someone puts the string `$123,456` into a field meant for numbers) and sometimes it's having too many or too few spaces before a bullet in a list.  When these things happen, the builds get weird.  If we're lucky, the build breaks and, thanks to a required check on the repo, it can't be pushed to production.  Other times, we're less fortunate and something gets through but functions incorrectly when the end-users engage with the site.

So, somewhere between my pedantic nature and my desire for to limit the impact of specification non-conformance on my non-work time, being able to lint the YAML front matter and verify it against a schema is a thing for me.

I appreciate your tolerating me and my questions and your above-and-beyond efforts to track down these issues I'm seeing are just amazing.  Thank you so very, very much.

**Comment by chris48s** [comment]
OK, so I've been on a bit of a voyage of discovery. I'm going to write up some proper documentation on this in the v8r repo, but here's a quick summary of what I've learned.

## Local Installs

As we've already established, there is no issue with using plugins (be they NPM packages or local files) if you have a local install of v8r and everything is in your `package.json`. I won't write any more about this.

## Global Install

If you have a global install (`npm install -g v8r`):

You can also install NPM package plugins into your global environment. So for example:

```
npm install -g v8r v8r-plugin-ndjson
v8r --format ndjson
```

works.

If you want to use local file plugins with a global install, the secret sauce is to run `npm link v8r` in your project dir. This will create a symlink in your project `node_modules` dir to your global `v8r` install. That will then allow you to `import ... from "v8r";` in local plugins.

## NPX

If you invoke v8r ad-hoc with npx, there is no way to get a local file plugin to work. However, you can also install plugin packages into the temporary environment. So for example,

```
npx --package v8r-plugin-ndjson@latest --package v8r@latest -- v8r --format ndjson
```

will work. That allows you to use plugins published as NPM packages. For local file plugins, you need an installation of some description.

## MegaLinter

MegaLinter effectively gives you a global install, but it lives in `/node-deps`.

Similar to a standard global install, you can install NPM packages into the global environment. So this MegaLinter config works:

```yml
ENABLE_LINTERS:
  - JSON_V8R
JSON_V8R_CLI_LINT_MODE: project
JSON_V8R_PRE_COMMANDS:
  - command: 'npm install v8r-plugin-ndjson'
    continue_if_failed: False
    cwd: root
JSON_V8R_ARGUMENTS: '--format ndjson'
```

The trick to getting local file plugins to work is similar to global installs, but `npm link` doesn't work in this case. I guess `npm link` doesn't know to look at the packages installed in `/node-deps`. However, the workaround is to create the symlink we need manually. This MegaLinter config does allow you to use local file plugins with MegaLinter:

```yml
ENABLE_LINTERS:
  - JSON_V8R
JSON_V8R_CLI_LINT_MODE: project
JSON_V8R_PRE_COMMANDS:
  - command: 'mkdir -p node_modules'
    continue_if_failed: False
    cwd: workspace
  - command: 'ln -s /node-deps/node_modules/v8r node_modules/v8r'
    continue_if_failed: False
    cwd: workspace
```

Note where I am using `cwd: root` and `cwd: workspace` in these two examples.

## Summary

It is possible to use plugins published as NPM packages with:

- Local install
- Global install `npm install -g v8r`
- Ad-hoc run `npx v8r@latest`
- Invocation via MegaLinter

It is possible to use local file plugins with:

- Local install
- Global install `npm install -g v8r`
- Invocation via MegaLinter

although some workarounds may be required.

## Other Notes

What _doesn't_ work is the `NODE_PATH` env var. This used to be a thing with CommonJS, but it no longer works with ES Modules. See https://github.com/orgs/nodejs/discussions/54916 for more info. In general, I tend to agree with the sentiment expressed in https://github.com/orgs/nodejs/discussions/54916#discussioncomment-10643665 If you're using and developing plugins, you probably want to explicitly pin your package versions and co-ordinate your plugin versions. With the MegaLinter model, at some point it is going to unexpectedly opt you in to breaking changes. In the event of breaking changes to the plugin interface, it is going to be fiddly to manage.

**Comment by chris48s** [comment]
I've now updated https://chris48s.github.io/v8r/plugins/using-plugins/ to cover this information.

I'm going to close this issue, but feel free to follow up if needed when you come back to this later in January.

**Comment by wesley-dean** [comment]
@chris48s I really appreciate your hard work on all of this -- it's way, way, way more than I expected!  I also apologize for not getting back to you sooner as it's been a crazy ride over here (U.S.) and my everything was up-ended in late January.  

Life circumstances being what they are, I recently came into some unanticipated availability which I'm using to write a few books, one of which is about OpenSSF's Scorecard and Allstar tools.  In that book, I've a section about validating Allstar configuration files (YAML) and I specifically call out `v8r` as _the tool_ to use for it.  I've also put together schema for the JSON Schema Store and I included in the repo validation scripts that use `v8r` to validate the schema files.

I've recently come to realize that a surprising amount of my talks, presentations, etc. that I've given have had an impact on people I didn't anticipate, often to people I didn't even know.  I'm saying that not to praise myself; rather, I hope to convey how something we do on a small scale can ripple out and affect others in profound ways.

It's like that for me with `v8r`.  I've been using it for quite a while but it wasn't until the past few months that I communicated that to anyone.  Seeing your good work and dedication to things like validating YAML and JSON files has been inspirational to me and I judge that I've learned a lot just by reading your documentation, studying the `v8r` repository, etc..

tl;dr: thank you, you're awesome
